File: src/main/scala/io/github/scala_tessella/editor/models/MenuShortcuts.scala
package io.github.scala_tessella.editor.models

/** Single source of truth for menu accelerator labels.
  *
  * Both [[io.github.scala_tessella.editor.components.MenuBarComponent]] (DOM menu hints) and the Tauri native
  * menu read from this table so the two can't drift; parity between this file and
  * `desktop/src-tauri/src/menu_shortcuts.rs` is enforced by the `checkMenuShortcutsParity` sbt task. Adding a
  * new labelled shortcut is one edit here; callers use [[MenuShortcuts.labelOf]].
  *
  * `primary` is the OS-conventional command modifier — Ctrl on Windows/Linux, Cmd on macOS. Label rendering
  * on the web always emits "Ctrl" (matches the pre-extraction behaviour); the Tauri menu in slice 2 renders
  * it as Cmd on macOS via its accelerator API rather than as a label string.
  */
object MenuShortcuts:

  enum MenuAction:
    case FileSave,
      EditUndo,
      EditRedo,
      EditDoubleToInfinite,
      EditDeselectAll,
      ViewFitToCanvas,
      ViewZoomIn,
      ViewZoomOut,
      ViewRotateLeft,
      ViewRotateRight

  case class Shortcut(
      key: String,
      primary: Boolean = false,
      shift: Boolean = false,
      alt: Boolean = false
  ):
    /** Renders "[Shift+][Ctrl+][Alt+]key". Shift-first ordering preserves the pre-extraction label format
      * (e.g. "Shift+Ctrl+Z" for redo), which is unusual vs. the common "Ctrl+Shift+Z" but changing it here
      * would be a user-visible regression.
      */
    def label: String =
      val parts =
        (if shift then Vector("Shift") else Vector.empty) ++
          (if primary then Vector("Ctrl") else Vector.empty) ++
          (if alt then Vector("Alt") else Vector.empty) ++
          Vector(key)
      parts.mkString("+")

  val bindings: Map[MenuAction, Shortcut] = Map(
    MenuAction.FileSave             -> Shortcut("S", primary = true),
    MenuAction.EditUndo             -> Shortcut("Z", primary = true),
    MenuAction.EditRedo             -> Shortcut("Z", primary = true, shift = true),
    MenuAction.EditDoubleToInfinite -> Shortcut("D"),
    MenuAction.EditDeselectAll      -> Shortcut("Esc"),
    MenuAction.ViewFitToCanvas      -> Shortcut("F"),
    MenuAction.ViewZoomIn           -> Shortcut("+"),
    MenuAction.ViewZoomOut          -> Shortcut("-"),
    MenuAction.ViewRotateLeft       -> Shortcut("E"),
    MenuAction.ViewRotateRight      -> Shortcut("R")
  )

  def of(action: MenuAction): Shortcut = bindings(action)

  def labelOf(action: MenuAction): String = of(action).label

File: desktop/src-tauri/src/menu_shortcuts.rs
//! Rust mirror of src/main/scala/.../models/MenuShortcuts.scala.
//!
//! The Scala side owns the DOM menu's display labels (Ctrl+S etc.).
//! This side owns the native menu's OS accelerators, which Tauri
//! remaps per-OS — "CmdOrCtrl" becomes ⌘ on macOS, Ctrl elsewhere.
//! Constant names here are the SCREAMING_SNAKE_CASE of the Scala
//! MenuAction enum values; MenuShortcutsParitySpec enforces that
//! every Scala case has a matching const here.

#![allow(dead_code)]

pub struct Shortcut {
    pub key: &'static str,
    pub primary: bool,
    pub shift: bool,
    pub alt: bool,
}

impl Shortcut {
    const fn key(key: &'static str) -> Self {
        Self { key, primary: false, shift: false, alt: false }
    }

    const fn primary(key: &'static str) -> Self {
        Self { key, primary: true, shift: false, alt: false }
    }

    const fn primary_shift(key: &'static str) -> Self {
        Self { key, primary: true, shift: true, alt: false }
    }

    /// Tauri accelerator string — "CmdOrCtrl+Shift+Z", "F", etc.
    pub fn accelerator(&self) -> String {
        let mut parts: Vec<&str> = Vec::new();
        if self.primary { parts.push("CmdOrCtrl"); }
        if self.shift   { parts.push("Shift"); }
        if self.alt     { parts.push("Alt"); }
        parts.push(self.key);
        parts.join("+")
    }
}

pub const FILE_SAVE:               Shortcut = Shortcut::primary("S");
pub const EDIT_UNDO:               Shortcut = Shortcut::primary("Z");
pub const EDIT_REDO:               Shortcut = Shortcut::primary_shift("Z");
pub const EDIT_DESELECT_ALL:       Shortcut = Shortcut::key("Escape");
pub const EDIT_ADD_COPY_TRANSLATE: Shortcut = Shortcut::key("T");
pub const EDIT_ADD_COPY_ROTATE:    Shortcut = Shortcut::key("R");
pub const EDIT_ADD_COPY_REFLECT:   Shortcut = Shortcut::key("Y");
pub const VIEW_FIT_TO_CANVAS:      Shortcut = Shortcut::key("F");
pub const VIEW_ZOOM_IN:            Shortcut = Shortcut::key("+");
pub const VIEW_ZOOM_OUT:           Shortcut = Shortcut::key("-");
pub const VIEW_ROTATE_LEFT:        Shortcut = Shortcut::key("Q");
pub const VIEW_ROTATE_RIGHT:       Shortcut = Shortcut::key("E");
